build-tools/packages/build-tools/src/common/gitRepo.ts

```diff
@@ -208,21 +208,39 @@ export class GitRepo {
 	 */
 	public async getModifiedFiles(): Promise<string[]> {
 		const results = await this.exec(`status --porcelain`, `get modified files`);
-		return results
-			.split("\n")
-			.filter((t) => t !== "" && !t.startsWith(" D "))
-			.map((t) => t.substring(3));
+		return (
+			results
+				.split("\n")
+				// Filter out empty lines and files that are deleted locally.
+				// Deleted files are marked with D in the first (staged) or second (unstaged) column.
+				// If a file is deleted in staged and then revived in unstaged, it will have two entries.
+				// The first entry will be "D " and the second entry will be "??". The second entry is
+				// will be enough to keep it.
+				.filter((t) => t !== "" && !t.match(/^.?D /))
+				.map((t) => t.substring(3))
+		);
 	}
 
 	/**
-	 * Returns an array containing repo root-relative paths to files that are deleted in the working tree.
+	 * Returns a set containing repo root-relative paths to files that are deleted in the working tree.
 	 */
-	public async getDeletedFiles(): Promise<string[]> {
+	public async getDeletedFiles(): Promise<Set<string>> {
```

Comment:
I'm a bit confused by the one call site for this function (which is in this file, so it probably shouldn't be public).

My understanding of how we use this git support is to detect changes to the repo to implement incremental build.
`getFiles` removes the deleted files from its changed file set, and thats the only use of `getDeletedFiles`.

This makes me think that deleting files could cause incremental build issues where we don't rebut when thats the only change.

If this is the case, this is the wrong fix, and we should not remove the deleted files. Instead we should include them, delete this method and make the incremental build logic correctly when  file goes from existing to deleted.

Reply:
It could be changed to private. The intent is to support an API set of utility functions. I would be surprised if this is the only one.

This is not the code that determines incrementality, but a set of low-level helper functions. (Making this private could hinder anything that needs the full set of changes to repo.)

But incremental build does not use (should not use) what is changed relative to what is committed. If we did that, then we've be up to date just switching branches.

There is some task (looks like `FlubCheckPolicyTask`) in full build that is specifically checking for modifications. I do not know the history. I expect that one task's incrementality is incorrect with respect to traditional incrementality. That is a different issue than being addressed here. (You are correct that if a file were deleted from an otherwise clean enlistment, then policy check would not be invoked. And a PR build would fail if the deletion were in violation of policy. Different issue; guessing quite low in priority.)

Reply:
The only use of getDeletedFiles is in:

```
	/**
	 * Returns an array containing repo repo-relative paths to all the files in the provided directory.
	 * A given path will only be included once in the array; that is, there will be no duplicate paths.
	 * Note that this function excludes files that are deleted locally whether the deletion is staged or not.
	 *
	 * @param directory - A directory to filter the results by. Only files under this directory will be returned. To
	 * return all files in the repo use the value `"."`.
	 */
	public async getFiles(directory: string): Promise<string[]> {
```

This seems like a really odd function. Based on its description, its just supposed to list the files in a directory. Nothing in its signature or comment implies this should only be modified files, and you can exclude deleted files by just not using git at all and just listing the directory's files, I don't understand why we are making two separate calls to `git` to implement `ls`.

I suppose implicitly this might be listing files which are not git ignored? Or maybe excluding ones which are added but not staged?

Personally, I don't think we should be writing a utility to provide some handy abstractions around git repos as part of FF's build tools. We should provide exactly what we need, and if we want general git things we should just use git, or some existing abstraction/library for it,

If we do however want to abstract git, we need to be way clearer with our abstraction, and how it relates to uncommitted, staged, ignored, etc. files. Right now we have utilities with a single use, and insufficiently documented to tell if the bug is in the implementation or user. 


"getFiles" seems to have a single use, `getBiomeFormattedFiles` (which computes it even when not needed), and runs it through an ignore filter ( https://www.npmjs.com/package/ignore) using a file ignore filtering library which can apply git ignore rules. Seems overly complex to ask git to list all files it knows about, then list modified files so you can filter out all the ones which are deleted, then run that through `ignore` to get a list of non-ignored files. I think a simple recursive directory walk using the `ignore` library and no git calls should work fine for that?

Anyway, given that deleted files can't fail lint, unless we are caching failed states, there shouldn't be an incremental build issue from that.

Reply:
> "getFiles" seems to have a single use, `getBiomeFormattedFiles` (which computes it even when not needed), and runs it through an ignore filter ( [npmjs.com/package/ignore](https://www.npmjs.com/package/ignore)) using a file ignore filtering library which can apply git ignore rules. Seems overly complex to ask git to list all files it knows about, then list modified files so you can filter out all the ones which are deleted, then run that through `ignore` to get a list of non-ignored files. I think a simple recursive directory walk using the `ignore` library and no git calls should work fine for that?

getBiomeFormattedFiles isn't using the ignore library to apply .gitignore files - it's using it to apply ignore patterns from the Biome config. I think you're right that we could instead walk the directories applying both .gitignore files and the biome config settings, but that seems like a lot more work than getting git itself to apply the .gitignores and then further filter that list based on Biome ignores.

> This seems like a really odd function. Based on its description, its just supposed to list the files in a directory. Nothing in its signature or comment implies this should only be modified files, and you can exclude deleted files by just not using git at all and just listing the directory's files, I don't understand why we are making two separate calls to git to implement ls.

I think one missing piece of context is that this function, getFiles, is in a class called GitRepo, so the description makes some assumptions that it's returning files git knows about. Perhaps `getTrackedFiles` `listFiles` would be a better name.